feat: Slack alert when VMs are on unapproved portgroups (#813)
vmware-samples/vcenter-event-broker-appliance: examples/knative/powercli/kn-pcli-pg-check/README.md

environment. You do not have to configure multiple portgroups, nor a mix of standard and
distributed portgroups.
This function was tested with the folowing configuration:

Tags
- New Tag Category named `PCI`
- New Tag named `PCI-VM` in Category `PCI`
- New Tag named `PCI-VM2` in Category `PCI`
- New Tag named `PCI-Network` in Category `PCI`

Port Groups
- New Distributed Virtual Portgroup named `DVS_PG_PCI_01`, tagged with `PCI/PCI-Network`
- New Distributed Virtual Portgroup named `DVS_PG_PCI_02`, tagged with `PCI/PCI-Network`
- New Distributed Virtual Portgroup named `DVS_PG_NOT_PCI`, no vSphere tag
- Existing Standard Virtual Portgroup named `VSS_VM_VLAN_203`, no vSphere tag

Virtual Machine
- New Virtual Machine named `tinycore-2`, tagged with `PCI/PCI-VM`
- Network Adapter 1 assigned to `DVS_PG_PCI_01`
- Newtork Adapter 2 assigned to `VSS_VM_VLAN_203`
- Network Adapter 3 assigned to `DVS_PG_NOT_PCI`
## Test your container image

Verify the container image works by executing it locally.

Change into the `test` directory
```console
cd test
```

Update the following variable names within the `docker-test-env-variable` file.

> > Note - The sample variables are built for a vCenter function. You will need to replace them if you are authoring a function for a different purpose.

Reviewer: looks like leftover from a template?

Author: Oops, deleted

> Any changes to variables must also be updated in `function_secret.yaml` and `test/docker-test-env-variable`

* `VCENTER_SERVER` - IP Address or FQDN of the vCenter Server to connect to
* `VCENTER_USERNAME` - vCenter account with permission to reconfigure distributed virtual switches
* `VCENTER_PASSWORD` - vCenter password associated with the username
* `VCENTER_CERTIFCATE_ACTION` - Set-PowerCLIConfiguration Action to configure when connection fails due to certificate error, default is Fail. (Possible values: Fail, Ignore or Warn)

> * `VM_WATCH_TAGS` - vCenter tags indicating a VM belongs on the PCI segment. In this example, `PCI/PCI-VM` and `PCI/PCI-VM2`.

Reviewer: question: are these variables then array/list types?

Author: Yes, they then become JSON arrays

> * `PG_WATCH_TAGS` - vCenter tags indicating a portgroup is a valid PCI segment. Int his example, `PCI/PCI-Network` and `PCI/PCI-Network2`

Reviewer: ```suggestion
* `PG_WATCH_TAGS` - vCenter tags indicating a portgroup is a valid PCI segment. In this example, `PCI/PCI-Network` and `PCI/PCI-Network2`
```

Author: Fixed

> you can chain all of the commands together with ampersands. This will allow you to re-run
> the commands by simply pressing the `up` arrow and `Enter`.

```console
cd .. && docker build -t ${IMAGE} . && cd test && docker run -e FUNCTION_DEBUG=true -e PORT=8080 --env-file docker-test-env-variable -it --rm -p 8080:8080 ${IMAGE}
```
# Step 3 - Deploy

> **Note:** The following steps assume a working Knative environment using the
`default` Rabbit `broker`. The Knative `service` and `trigger` will be installed in the
`vmware-functions` Kubernetes namespace, assuming that the `broker` is also available there.

If you built a custom image, push it to an accessible registry such as Docker once you're done developing and testing your function logic.

```console
docker push ${IMAGE}
```

Update the `function_secret.json` file with the same vCenter Server credentials and configurations that you used in `test/docker-test-env-variable` and then create the kubernetes secret which can then be accessed from within the function by using the environment variable named called `FUNCTION_SECRET`.

```console
# create secret

> 

Reviewer: nit: redundant newline

Author: Fixed

kubectl -n vmware-functions create secret generic pg-check-secret --from-file=FUNCTION_SECRET=pg_check_secret.json

# update label for secret to show up in VEBA UI
kubectl -n vmware-functions label secret pg-check-secret app=veba-ui
```

Edit the `function.yaml` file with the name of the container image from Step 1 if you made any changes. If not, the default VMware container image will suffice. By default, the function deployment will filter on the `VmReconfiguredEvent` vCenter Server Event. If you wish to change this, update the `subject` field within `function.yaml` to the desired event type.

Deploy the function to the VMware Event Broker Appliance (VEBA).

```console
# deploy function

> 

Reviewer: nit: redundant newline

Author: fixed
